File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章字符串转换.py
#编辑距离， 字符串转换 lanqiaoOJ 1507


a = input(); a=' '+a  #a[0]不用，用a[1]~a[m]
b = input(); b=' '+b
m = len(a)-1
n = len(b)-1
dp = [[0] * (n + 1) for _ in range(m + 1)]
for i in range(1,m+1 ):    dp[i][0] = i
for i in range(1,n+1 ):    dp[0][i] = i
for i in range(1,m+1):
    for j in range(1,n+1):
        if a[i] == b[j]:  dp[i][j] = dp[i-1][j-1]
        else:  dp[i][j]=min(min(dp[i-1][j],dp[i][j-1]),dp[i-1][j-1])+1
print(dp[m][n])

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章字符串转换.java
//编辑距离， 字符串转换 lanqiaoOJ 1507
import java.util.Scanner;


public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s1=scanner.nextLine();
        String s2=scanner.nextLine();
        int[][] dp=new int[s1.length()+1][s2.length()+1];
        for(int i = 1 ; i <= s1.length() ; i++ ){  //记得初始化数据
            dp[i][0] = i ;
        }
        for(int i = 1 ; i <= s2.length() ; i++ ){
            dp[0][i] = i ;
        }
        for (int i = 1; i <=s1.length(); i++) {
            for (int j =1; j <=s2.length(); j++) {
                if (s1.charAt(i-1)==s2.charAt(j-1)){
                    dp[i][j]=dp[i-1][j-1];
                }else{
                    dp[i][j]=Math.min(Math.min(dp[i-1][j-1]+1,dp[i-1][j]+1),dp[i][j-1]+1);
                }
            }
        }
        System.out.println(dp[s1.length()][s2.length()]);
    }
}